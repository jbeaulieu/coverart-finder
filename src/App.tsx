import { createRef, useState } from 'react'
import viteLogo from '/vite.svg'
import './App.css'
import SearchContainer from './components/SearchContainer'
import type { Album } from './@types/album'
import AlbumListContainer from './components/AlbumListContainer'
import SizeSlider from './components/SizeSlider'
import CoverPreviewContainer from './components/CoverPreviewContainer'
import ITunesApiClient from './util/api/itunesApiClient'
import { getITunesArtworkUrl } from './util/url/urlParser'
import type { FixedSizeList } from 'react-window'
import { Card, CardContent, Typography } from '@mui/material'

const album: Album = {
  id: 1767673630,
  name: 'Watch the Fire',
  artistName: 'Boys Like Girls',
  thumbnailSrc: "https://is1-ssl.mzstatic.com/image/thumb/Music221/v4/c0/4b/01/c04b0132-0f23-2a9a-88f4-c22d996470e9/067003846363.png/60x60bb.jpg"
};

const album2: Album = {
  id: 276649794,
  name: 'Boys Like Girls',
  artistName: 'Boys Like Girls',
  thumbnailSrc: "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/04/5a/e4/045ae4df-461d-db63-9b6f-cb48507c682d/mzi.ddfuhxng.jpg/60x60bb.jpg"
}

const album3: Album = {
  id: 580265083,
  name: 'Crazy World',
  artistName: 'Boys Like Girls',
  thumbnailSrc: "https://is1-ssl.mzstatic.com/image/thumb/Music124/v4/d9/bd/65/d9bd655f-7de6-30a1-c1dc-9db071e9879d/886443665224.jpg/60x60bb.jpg"
}

function App() {
  const [list, setList] = useState([album, album2, album3]);
  const [selectedAlbum, setSelectedAlbum] = useState<Album>(album);
  const [imageSize, setImageSize] = useState(1000);

  const listRef = createRef<FixedSizeList>();

  const doSearch = async (query: string) => {
    try {
      const results = await ITunesApiClient.searchAlbums(query);
      setList(results);
    } catch (error) {
      console.error(`Error while fetching results: ${error}`);
    }

    // Scroll to top after setting new search results
    listRef.current?.scrollToItem(0);
  };

  const updateSelected = (selectedId: number) => {
    const selection = list.find((a) => a.id == selectedId);
    setSelectedAlbum(selection!);
  };

  const getPreviewPaneArtwork = () => {
    if (!selectedAlbum.thumbnailSrc) return "";
    return getITunesArtworkUrl(selectedAlbum.thumbnailSrc, 600);
  };

  const getRequestedSizeArtwork = () => {
    if (!selectedAlbum.thumbnailSrc) return "";
    return getITunesArtworkUrl(selectedAlbum.thumbnailSrc, imageSize);
  };

  return (
    <>
      <div>
        <a href="https://vite.dev" target="_blank">
          <img src={viteLogo} className="logo" alt="Vite logo" />
        </a>
      </div>
      <Card style={{ padding: 50 }}>
        <CardContent>
          <Typography fontSize={32} marginBottom={4} sx={{ color: 'text.secondary' }}>
            Coverart-Finder
          </Typography>
          <div id="container" style={{ display: 'flex', gap: 60 }}>
            <div id="finder-container" style={{ display: 'flex', flexDirection: 'column', alignItems: 'start' }}>
              <SearchContainer doSearch={(input: string) => doSearch(input)} />
              <AlbumListContainer albumList={list} listRef={listRef} selectedAlbumId={selectedAlbum.id} onSelect={(id) => updateSelected(id)} />
              <SizeSlider selectedSize={imageSize} setSelectedSize={setImageSize} />
            </div>
            <CoverPreviewContainer selectedSize={imageSize} previewSrc={getPreviewPaneArtwork()} downloadSrc={getRequestedSizeArtwork()} />
          </div>
        </CardContent>
      </Card>
      <p className="read-the-docs">
        Click on the Vite and React logos to learn more
      </p>
    </>
  )
}

export default App;
