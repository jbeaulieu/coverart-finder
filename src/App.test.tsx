import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import App from './App';

const renderApp = () => {
  return render(<App />);
};

describe('Component App', () => {
  it('renders without crashing', () => {
    const { asFragment } = renderApp();

    expect(asFragment()).toMatchSnapshot();
  });
});
