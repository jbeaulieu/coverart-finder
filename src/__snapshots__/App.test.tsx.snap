// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component App > renders without crashing 1`] = `
<DocumentFragment>
  <img
    alt="App logo"
    class="logo"
    src="/src/assets/icon.png"
  />
  <p
    class="MuiTypography-root MuiTypography-body1 css-5c82k1-MuiTypography-root"
  >
    Coverart Finder
  </p>
  <span
    class="MuiSwitch-root MuiSwitch-sizeMedium css-rehya9-MuiSwitch-root"
  >
    <span
      class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-aqlqb4-MuiButtonBase-root-MuiSwitchBase-root-MuiSwitch-switchBase"
    >
      <input
        class="PrivateSwitchBase-input MuiSwitch-input css-12xagqm-MuiSwitchBase-root"
        type="checkbox"
        value="false"
      />
      <span
        class="MuiSwitch-thumb css-17jyosd-MuiSwitch-thumb"
      />
    </span>
    <span
      class="MuiSwitch-track css-tj8d99-MuiSwitch-track"
    />
  </span>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-152icmn-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      id="finder-container"
      style="display: flex; flex-direction: column; align-items: start;"
    >
      <form
        autocomplete="off"
        class="MuiBox-root css-7wz4ac"
        novalidate=""
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="search-container"
            id="search-container-label"
          >
            Search by Artist / Album
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-1n04w30-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
              id="search-container"
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
            >
              <button
                aria-label="search"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1ysp02-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="submit"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="SearchIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                  />
                </svg>
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1n64csd-MuiNotchedOutlined-root"
              >
                <span>
                  Search by Artist / Album
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </form>
      <div
        class="MuiBox-root css-15v79bh"
      >
        <div
          style="position: relative; height: 500px; width: 500px; overflow: auto; -webkit-overflow-scrolling: touch; will-change: transform; direction: ltr;"
        >
          <div
            style="height: 216px; width: 100%;"
          >
            <li
              class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
              style="position: absolute; left: 0px; top: 0px; height: 72px; width: 100%;"
            >
              <div
                class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiListItemButton-root MuiListItemButton-gutters Mui-selected css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                  >
                    Watch the Fire
                  </span>
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
                  >
                    Boys Like Girls
                  </p>
                </div>
                <div
                  class="MuiListItemAvatar-root css-1ulznza-MuiListItemAvatar-root"
                >
                  <div
                    class="MuiAvatar-root MuiAvatar-circular css-fmeqll-MuiAvatar-root"
                  >
                    <img
                      alt="Watch the Fire"
                      class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                      src="https://is1-ssl.mzstatic.com/image/thumb/Music221/v4/c0/4b/01/c04b0132-0f23-2a9a-88f4-c22d996470e9/067003846363.png/60x60bb.jpg"
                    />
                  </div>
                </div>
              </div>
            </li>
            <li
              class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
              style="position: absolute; left: 0px; top: 72px; height: 72px; width: 100%;"
            >
              <div
                class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                  >
                    Boys Like Girls
                  </span>
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
                  >
                    Boys Like Girls
                  </p>
                </div>
                <div
                  class="MuiListItemAvatar-root css-1ulznza-MuiListItemAvatar-root"
                >
                  <div
                    class="MuiAvatar-root MuiAvatar-circular css-fmeqll-MuiAvatar-root"
                  >
                    <img
                      alt="Boys Like Girls"
                      class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                      src="https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/04/5a/e4/045ae4df-461d-db63-9b6f-cb48507c682d/mzi.ddfuhxng.jpg/60x60bb.jpg"
                    />
                  </div>
                </div>
              </div>
            </li>
            <li
              class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
              style="position: absolute; left: 0px; top: 144px; height: 72px; width: 100%;"
            >
              <div
                class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                  >
                    Crazy World
                  </span>
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
                  >
                    Boys Like Girls
                  </p>
                </div>
                <div
                  class="MuiListItemAvatar-root css-1ulznza-MuiListItemAvatar-root"
                >
                  <div
                    class="MuiAvatar-root MuiAvatar-circular css-fmeqll-MuiAvatar-root"
                  >
                    <img
                      alt="Crazy World"
                      class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                      src="https://is1-ssl.mzstatic.com/image/thumb/Music124/v4/d9/bd/65/d9bd655f-7de6-30a1-c1dc-9db071e9879d/886443665224.jpg/60x60bb.jpg"
                    />
                  </div>
                </div>
              </div>
            </li>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-qxh7fn"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-vjrxpc-MuiTypography-root"
          id="input-slider"
        >
          Download Size (px)
        </p>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-1bcqwu3-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-grow css-1ynczkt-MuiGrid-root"
          >
            <span
              class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-sizeMedium css-13vpke-MuiSlider-root"
            >
              <span
                class="MuiSlider-rail css-r64h58-MuiSlider-rail"
              />
              <span
                class="MuiSlider-track css-xvk2i-MuiSlider-track"
                style="left: 0%; width: 100%;"
              />
              <span
                class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-1xcmt9q-MuiSlider-thumb"
                data-index="0"
                style="left: 100%;"
              >
                <input
                  aria-labelledby="input-slider"
                  aria-orientation="horizontal"
                  aria-valuemax="1000"
                  aria-valuemin="50"
                  aria-valuenow="1000"
                  data-index="0"
                  max="1000"
                  min="50"
                  step="50"
                  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                  type="range"
                  value="1000"
                />
              </span>
            </span>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1lair6c-MuiTypography-root"
            style="margin-bottom: 4px;"
          >
            1000
          </p>
        </div>
      </div>
    </div>
    <div
      class="preview-container"
    >
      <img
        alt="cover art preview"
        height="600"
        src="https://is1-ssl.mzstatic.com/image/thumb/Music221/v4/c0/4b/01/c04b0132-0f23-2a9a-88f4-c22d996470e9/067003846363.png/600x600bb.jpg"
        style="object-fit: cover;"
        width="600"
      />
      <div
        style="display: flex; gap: 20px; place-items: center;"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Download
          <span
            class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-1gulhci-MuiButton-endIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="DownloadIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5 20h14v-2H5zM19 9h-4V3H9v6H5l7 7z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
